### PR TITLE
Introduce GoogleTest framework for Valkey unit testing

### PR DESCRIPTION
**Pull request by harrylin98**
### This PR adds GoogleTest (gtest) support to Valkey to enable writing modern unit tests.

**Motivation**: 
GoogleTest provides richer assertions, test fixtures, mocking support, and improved diagnostics, helping improve test coverage and maintainability over time.

**Summary**:
The change is limited to test infrastructure, and existing C unit tests remain unchanged.

This PR focuses only on integrating the framework and includes a small set of example tests to demonstrate usage.

For more details, see `src/gtest/README.md`.

### CODE REVIEW

**Review by JimB123 — APPROVED** [review]
I like the documentation and the integration with the current unit tests.  We can see the example tests already running as the test builds are executed.  Nice work pulling this all together and ensuring support across all of the platforms.

This provides the capability for much richer unit testing.

I look forward to using this during the development of the forkless capabilities: https://github.com/valkey-io/valkey/issues/2878

**Comment by madolson** [comment]
Reposting my top level comment:s

* I don't think we should have two unit testing frameworks. I feel like we need to commit to incorporating the google testing framework. I'm not saying that is a blocker to start, but we should agree to move what we have.
* We typically deep copy dependencies to make vending easier, maybe we don't care in this case because it is optional and not required? Not a particularly strong opinion.

File: src/gtest/README.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+- Integrated mocking via GoogleMock, providing expressive, behavior-based mocks
+- Rich argument matchers and call sequencing / ordering
+- Advanced test fixtures and lifecycle control
+- Richful assertions and detailed diagnostics
+- Parameterized and typed tests
+
+These features enable more expressive, maintainable, and scalable unit tests,
+particularly for complex components and edge-case validation.
+
+For more information on GoogleTest, see: https://google.github.io/googletest/
```
Comment (madolson) [inline src/gtest/README.md]:
```suggestion
```

Still recommend cutting this. It's not really relevant, most of these are relative or normative statements. I would move the link to the google test lower.

File: src/gtest/example_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+// Test matcher works in custom_matchers.hpp
+TEST_F(ExampleTest, TestMatchers) {
+    robj *robj_str = createStringObject("test", 4);
+    ASSERT_NE(robj_str , nullptr); // "ASSERT" is correct here, because the test can't reasonably continue
+    EXPECT_THAT(robj_str, robjEqualsStr("test"));
+    decrRefCount(robj_str);
+}
+
+// Verify mocking works via zfree
+TEST_F(ExampleTest, TestMocking) {
```
Comment (madolson) [inline src/gtest/example_tests.cpp:61]: I think this is the type of mocking that got ElastiCache into trouble. This is trying to make sure zfree is called, but there are better systemic ways like deprecating not using free (See https://github.com/valkey-io/valkey/blob/0a9cb985c8c37c54c5007bc0af2dcd0b6752a529/src/server.h#L4077)

File: src/gtest/wrappers.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+#define __WRAPPERS_H
+#define _Atomic /* nothing */
+#define _Bool bool
+#define typename _typename
+#define protected protected_
+#include "ae.h"
+#include "dict.h"
+#include "server.h"
+#include "adlist.h"
+#include "zmalloc.h"
```
Comment (madolson) [inline src/gtest/wrappers.h]: Do we need server.h here? I feel like a bunch of these are not useful.

File: src/gtest/wrappers.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ */
+
+#include <sched.h>
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef __WRAPPERS_H
+#define __WRAPPERS_H
+#define _Atomic /* nothing */
```
Comment (madolson) [inline src/gtest/wrappers.h]: What does the nothing comment mean?

File: src/gtest/wrappers.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ * The list of wrapper methods defined.  Each wrapper method must
+ * conform to the same naming conventions (i.e. prefix with a
+ * '__wrap_') and have its method signature match the overridden
+ * method exactly.
+ *
+ * Note: You should NOT wrap variable argument functions (i.e have "...")
+ *       See: https://github.com/google/googletest/blob/master/googlemock/docs/gmock_faq.md#can-i-mock-a-variadic-function
+ *       Example: serverLog(int level, const char *fmt, ...) should NOT be mocked.
+ */
+long long __wrap_aeCreateTimeEvent(aeEventLoop *eventLoop, long long milliseconds, aeTimeProc *proc, void *clientData, aeEventFinalizerProc *finalizerProc);
```
Comment (madolson) [inline src/gtest/wrappers.h:60]: I would like to just keep this one, and remove the other wraps. I don't think you should be mocking creation of basic objects. I think you want to be mocking things you don't get deterministic results from (like timers, network, disk etc.

File: src/Makefile [hunk 6]
```diff
@@ -482,7 +485,12 @@ endif
 
 .PHONY: all
 
+# gtest unit tests are not supported on macOS due to linker incompatibility
```
Comment (madolson) [inline src/Makefile]: I've used gtest on MacOS?

FWIW, I think this is a blocker. A good chunk of folks that work on valkey only use macOS, so this needs to be supported.

File: src/gtest/Makefile [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+	TEST_CFLAGS+=-DFORCE_RUN_SKIPPED_TESTS
+endif
+
+# Track compilation flags to force rebuild when they change
+.PHONY: .flags
+.flags:
+	@echo '$(TEST_CFLAGS)' | cmp -s - $@ || echo '$(TEST_CFLAGS)' > $@
+
+# OSS Valkey library dependencies
+VALKEY_LIB := ../libvalkey.a
```
Comment (madolson) [inline src/gtest/Makefile]: The Valkey client is called libvalkey. We build this as valkey.a in the top level directory. I don't know if we need to rebuild it here, and if we do, let's call it something else.

File: src/gtest/wrappers.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+ * 4. GoogleTest can mock these wrappers to control behavior and verify calls
+ *
+ * RULES:
+ * - All wrapper functions MUST be prefixed with __wrap_
+ * - Function signatures MUST exactly match the original C function
+ * - DO NOT wrap variadic functions (functions with ...) - GoogleTest doesn't support them
+ * - Each wrapper becomes mockable in gtest via the auto-generated MockValkey class
+ *
+ * WORKFLOW:
+ * wrappers.h -> generate-wrappers.py -> [generated_wrappers.hpp + generated_wrappers.cpp]
```
Comment (madolson) [inline src/gtest/wrappers.h:23]: These two files don't seem to be gitignored.

File: src/gtest/Makefile [hunk 9]
```diff
@@ -0,0 +1,178 @@
+CXX ?= g++
+-include ../.make-settings
+OPT=-O3 -DNDEBUG
+
+ifeq ($(DEBUG_BUILD),1)
```
Comment (madolson) [inline src/gtest/Makefile]: What is this DEBUG_BUILD option? EDIT: I think this might be an AWS thing.

File: src/gtest/example_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,66 @@
+#include "generated_wrappers.hpp"
+
+extern "C" {
+    #include "dict.h"
+}
+
+// Use a class name descriptive of your test unit
```
Comment (madolson) [inline src/gtest/example_tests.cpp:7]: I think it is required to end with TEST no? Would be useful to document any other requirements since this is the example.

File: src/gtest/example_tests.cpp
Comment (madolson) [inline src/gtest/example_tests.cpp:1]: We should add a style checker for CPP tests.

**Review by madolson — COMMENTED** [review]
I think there is a bit more cleanup to do in make files, but I think they're close as of right now.

Left some other comments.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -55,6 +58,11 @@ jobs:
           - {version: "8.1.4", file: "valkey-8.1.4-noble-x86_64.tar.gz"}
     steps:
       - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+      - name: Install modern CMake
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y cmake
+          cmake --version
```
Comment (madolson) [inline .github/workflows/ci.yml]:
```suggestion
```

**Comment by madolson** [comment]
@harrylhl Also check the 32 bit build, which seems to be failing now.

**Comment by zuiderkwast** [comment]
> * We typically deep copy dependencies to make vending easier, maybe we don't care in this case because it is optional and not required? Not a particularly strong opinion.

I prefer that we don't vendor it, since it's optional and only for developers, not useful for end users. I think we should handle it like we handle the optional OpenSSL, i.e. the user has to install it locally using `apt install` or similar.

**Comment by madolson** [comment]
> I prefer that we don't vendor it

That's a good point. It seems to available on hombrew and most most package managers. Let's do that!

File: src/gtest/wrappers.h (hunk 3 above)
Comment (harrylin98) [reply src/gtest/wrappers.h]: server.h pulls in core headers and ensures consistent C/C++ compatibility, so I’d suggest keeping it included.

File: src/gtest/wrappers.h (hunk 4 above)
Comment (harrylin98) [reply src/gtest/wrappers.h]: It means `_Atomic` is being defined as an empty macro.

This strips the `_Atomic` keyword so the code can compile as C++ (where _Atomic is not a valid keyword). This is a compatibility trick when sharing headers between C and C++.

File: src/gtest/wrappers.h (hunk 8 above)
Comment (harrylin98) [reply src/gtest/wrappers.h:23]: They are ignored in `src/.gitignore`

File: src/gtest/wrappers.h (hunk 4 above)
Comment (madolson) [reply src/gtest/wrappers.h]: Can you document that instead of a "trick".

File: src/gtest/wrappers.h (hunk 3 above)
Comment (madolson) [reply src/gtest/wrappers.h]: The main Valkey project doesn't guarantee C/C++ compatibility, it's a C project. I don't want to enforce that as a constraint.

File: src/gtest/Makefile (hunk 9 above)
Comment (harrylin98) [reply src/gtest/Makefile]: Will remove

File: src/gtest/example_tests.cpp (hunk 10 above)
Comment (harrylin98) [reply src/gtest/example_tests.cpp:7]: The class name itself can be anything, GoogleTest only requires the fixture inherits from `::testing::Test`

File: src/gtest/example_tests.cpp
Comment (harrylin98) [reply src/gtest/example_tests.cpp:1]: Can you specify the requirement.

Comment (madolson) [reply src/gtest/example_tests.cpp:1]: Add a style checker like we have for the C code. We use clang-format, see https://github.com/valkey-io/valkey/blob/unstable/src/.clang-format.

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -84,10 +91,12 @@ jobs:
       - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
       - name: cmake and make
         run: |
+          sudo apt-get update
           sudo apt-get install -y cmake libssl-dev
+          cmake --version
```
Comment (madolson) [inline .github/workflows/ci.yml]: Do we need to print out the version here?

File: deps/Makefile [hunk 13]
```diff
@@ -128,3 +129,27 @@ fast_float_c_interface: .make-prerequisites
 	cd fast_float_c_interface && $(MAKE)
 
 .PHONY: fast_float_c_interface
+
+CMAKE := $(shell command -v cmake3 || command -v cmake)
```
Comment (madolson) [inline deps/Makefile]: From viktor's earlier comment, let's install these instead of having them be vendored. They are not strictly required to build Valkey. I'm okay fully copying gtest-parallel, given how small it is though. I don't want to be building gtest though.

File: src/gtest/.clang-format [hunk 14]
```diff
@@ -0,0 +1,38 @@
+BasedOnStyle: LLVM
```
Comment (madolson) [inline src/gtest/.clang-format:1]: I don't see any place this is actually enforced, should probably be in the ci.yml

File: src/gtest/Makefile [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+	-isystem .. -I ../../deps/lua/src/ -I ../../deps/hdr_histogram/ -I ../../deps/fpconv/ $(JEMALLOC_CFLAGS) \
+	-I ../../deps/googletest/googletest/include -I ../../deps/googletest/googlemock/include $<
+
+$(ALL_OBJECTS): OPT := $(OPT) $(MORE_WARNINGS)
+
+# Obtain the list of function calls for which we are generating mocks using the gmock library.  We assume they are defined in
+# wrappers.h and begin with the prefix '__wrap_' (which is required by the linker).  These command line arguments are then
+# passed in to g++ during linking.
+ifneq ($(wildcard wrappers.h),)
+CMD:=egrep '__wrap_.*\(.*;' wrappers.h | sed 's/__wrap_/@/' | sed 's/[^@]*@\([^(]*\)(.*/\1/' | sort | uniq | xargs -I{} echo "-Wl,--wrap={}" | tr "\n" " "
```
Comment (madolson) [inline src/gtest/Makefile]: Kind of a nitpick, but sometimes you indent commands under if/ifneq and sometimes you don't. Would be nice to be consistent.

File: .github/workflows/ci.yml (hunk 12 above)
Comment (harrylin98) [reply .github/workflows/ci.yml]: Nope, will remove.

File: src/gtest/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+    endif ()
+endif ()
+
+# Fetch GoogleTest using FetchContent (modern approach)
+include(FetchContent)
+FetchContent_Declare(
+  googletest
+  URL https://github.com/google/googletest/archive/03597a01ee50ed33e9dfd640b249b4be3799d395.zip
+)
+# For Windows: Prevent overriding the parent project's compiler/linker settings
```
Comment (murphyjacob4) [inline src/gtest/CMakeLists.txt]: We don't support Windows builds, should we drop this?

File: src/gtest/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+        list(APPEND COMPILE_DEFINITIONS "USE_OPENSSL=1")
+        list(APPEND COMPILE_DEFINITIONS "BUILD_TLS_MODULE=0")
+    endif ()
+endif ()
+
+# Fetch GoogleTest using FetchContent (modern approach)
+include(FetchContent)
+FetchContent_Declare(
+  googletest
+  URL https://github.com/google/googletest/archive/03597a01ee50ed33e9dfd640b249b4be3799d395.zip
```
Comment (murphyjacob4) [inline src/gtest/CMakeLists.txt]: Don't we have a git submodule we can use? Why clone the source here

File: src/gtest/wrappers.h [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+ * The list of wrapper methods defined.  Each wrapper method must
+ * conform to the same naming conventions (i.e. prefix with a
+ * '__wrap_') and have its method signature match the overridden
+ * method exactly.
+ *
+ * Note: You should NOT wrap variable argument functions (i.e have "...")
+ *       See: https://github.com/google/googletest/blob/master/googlemock/docs/gmock_faq.md#can-i-mock-a-variadic-function
+ *       Example: serverLog(int level, const char *fmt, ...) should NOT be mocked.
+ */
+int __wrap_processCommand(struct client *c);
```
Comment (murphyjacob4) [inline src/gtest/wrappers.h]: I think the big problem here is that the majority of Valkey is just not really designed to be unit testable. To workaround this, it seems we are giving the ability for arbitrary function calls to be mocked. I'm worried about the fragility.

If everything can be mocked, it becomes natural (and fast) to write tests that stub out most of the world. Those tests often end up asserting something like "function A calls function B with argument X and then returns Y". In fact this is kind of like the test example you wrote.

More often than not this type of test is simply a description of the behavior, not a stable contract. A small refactor breaks tests even though the externally-correct behavior may be unchanged.

Normally, mocks are acceptable when you "mock at boundaries". This way, the boundary interfaces are stable and we don't end up with tests not compiling or passing after small changes, only when those boundaries change. In contrast, I feel this framework is effectively employing "mock everything".

Overall, I'm not opposed to mocking, and the wrap strategy seems to be somewhat effective compilation strategy to get mocking to work with in the C world, but I am worried about the effects this will have on maintainability. Mocking processCommand feels like a wrong precedent to set. I feel like we should be mocking those boundaries (things like clocks, filesystems, networks).

For new components (maybe BGIterator is one) - injecting function pointers/a `vtable`-like struct is also a viable strategy. Then the test can inject test fakes for those functions instead of mocks. At least there we define a clear boundary (i.e. BGIterator needs functions A, B, and C with these contracts)

**Comment by harrylin98** [comment]
> I prefer that we don't vendor it, since it's optional and only for developers, not useful for end users. I think we should handle it like we handle the optional OpenSSL, i.e. the user has to install it locally using `apt install` or similar.

Okay, i will remove GoogleTest from the Deps/.

File: deps/Makefile (hunk 13 above)
Comment (harrylin98) [reply deps/Makefile]: Agree, I will remove googleTest. No need to track/patch googletest vulnerabilities

File: src/gtest/CMakeLists.txt (hunk 16 above)
Comment (harrylin98) [reply src/gtest/CMakeLists.txt]: Sure, will remove.

File: src/gtest/CMakeLists.txt (hunk 17 above)
Comment (harrylin98) [reply src/gtest/CMakeLists.txt]: I will remove the googleTest submodules in next revision, users will need to use system-installed GoogleTest.
Will also remove the FetchContent sections here.

File: src/gtest/wrappers.h (hunk 18 above)
Comment (harrylin98) [reply src/gtest/wrappers.h]: I agree with the concern about fragility for mocking, and that mocks are best used at stable boundaries. The intent here is to make some otherwise-untestable logic tesable that lacks clear boundaries.

I think a reasonable way right now is to define guidelines on what is acceptable to mock.

Comment (JimB123) [reply src/gtest/wrappers.h]: Fundamentally, we all agree that mocking (like any technology) can be misused.  However, the ability to mock is critical to support isolation for unit testing.  Misuse is addressed in PR reviews.

>I think the big problem here is that the majority of Valkey is just not really designed to be unit testable.

It's true that much of the existing Valkey code is not designed well for unit testing.  There are areas without strong APIs.  There are areas that don't strongly adhere to the principals of low-coupling and high-cohesion.  And you're right; this may create a problem for some unit testing.

However there ARE areas of Valkey that do have well defined APIs (boundaries) and do provide examples of low-coupling and high-cohesion.  In order to test these areas, mocking is appropriate and well suited.

>Mocking processCommand feels like a wrong precedent to set.

Note that this is an arbitrary example - and maybe not the best.  But it's not inherently bad.  `processCommand` is as close as exists to an "API" (boundary) function for `server.c`.  It's called from `networking.c`, which is quite arguably a separate "unit".  If we want to unit test functionality in `networking.c` (which we don't do today), it would be very reasonable to mock `processCommand`.

I can also envision a use case where we are unit testing a blocking mechanism (separate "unit"), and want to ensure that `processCommand` gets called when a client is unblocked.  This would also be a legitimate use of such a mock.

>For new components

As Valkey continues to grow, it will become more critical to create better defined units, clearer APIs, and reduced coupling.

>BGIterator

Since you mentioned BgIterator, at Amazon, the bgIteration tests use a fairly limited set of mocks to provide isolation:
* block/unblock a client
* pause/unpause rehashing
* create a timer event
* get expiry time
* permission check a client command

That's it.  It's possible to perform unit testing without going crazy with the mocks.

Comment (zuiderkwast) [reply src/gtest/wrappers.h]: > `processCommand` is as close as exists to an "API" (boundary) function for `server.c`. It's called from `networking.c`, which is quite arguably a separate "unit". If we want to unit test functionality in `networking.c` (which we don't do today), it would be very reasonable to mock `processCommand`.

I don't want a unit test for `processCommand` as it's designed today. Either we stick to only integration testing for these parts or we need to redesign these APIs so they can be unit-tested without extensive mocking.

File: deps/gtest-parallel
Comment (zuiderkwast) [inline deps/gtest-parallel:1]: We're not using submodules for dependencies so far. If we vendor the code, we should follow the style used for jemalloc and libvalkey. For these, we have used git *subtree* rather than *submodule*. See https://stackoverflow.com/questions/32407634/when-to-use-git-subtree

Comment (harrylin98) [reply deps/gtest-parallel:1]: Right now, I’ve completely removed the GoogleTest submodule, and users are required to install GoogleTest on their system instead.

I’ve kept the gtest-parallel package as Maddy suggested. It is currently included as a submodule, but I will update it to follow the same approach we use for jemalloc and libvalkey (using a git subtree)

File: src/gtest/wrappers.h (hunk 18 above)
Comment (harrylin98) [reply src/gtest/wrappers.h]: I will replace with another example by using aeCreateTimeEvent.

File: deps/gtest-parallel
Comment (zuiderkwast) [reply deps/gtest-parallel:1]: Yeah, I've seen you've already removed gtest submodule. 👍

File: src/gtest/CMakeLists.txt (hunk 17 above)
Comment (harrylin98) [reply src/gtest/CMakeLists.txt]: Hey, I think here I would still keep the FetchContent.

1. GoogleTest submodule is removed from Deps/
2. We intentionally use CMake FetchContent instead of a git submodule or system GTest so that GoogleTest is built with the same compiler, flags, and link model as Valkey.
FetchContent guarantees GTest is compiled as part of this build and stays compatible.

Check [https://google.github.io/googletest/quickstart-cmake.html](https://google.github.io/googletest/quickstart-cmake.html)

File: deps/gtest-parallel
Comment (harrylin98) [reply deps/gtest-parallel:1]: Once I moved to git subtree, I got a bunch of missing sign-off commits from gtest-parallel. Any ideas of how to resolve this.

Comment (harrylin98) [reply deps/gtest-parallel:1]: https://github.com/google/gtest-parallel/issues/100
I sent a request to them to add the sign-off

**Comment by harrylin98** [comment]
Hi @madolson @zuiderkwast @murphyjacob4,

I believe the current version of the code is ready for another review. The remaining DCO failures are due to missing Signed-off-by lines in the gtest-parallel commits, which I am actively working to resolve.

When you have time, could you please take another look? Thanks!

File: src/gtest/README.md [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+and RealValkey.
+
+MockValkey uses gtest definitions to define a mock class. RealValkey uses the
+__real_foo() methods to call the renamed methods. The script also implements
+every __wrap_foo() command that delegates to the last MockValkey instance
+initialized.
+
+To extend the Valkey classes for mocking further methods, simply add your method
+to 'wrappers.h' and re-run 'make test-gtest' to regenerate the Valkey glue code
+and run the tests.
```
Comment (murphyjacob4) [inline src/gtest/README.md:25]: Can we add some guidelines about new mocking points?

```suggestion
To extend the Valkey classes for mocking further methods, simply add your method
to 'wrappers.h' and re-run 'make test-gtest' to regenerate the Valkey glue code
and run the tests.

Important: All mocking should occur at software boundaries where interfaces are
clearly defined. Your use of mocking will be denied if it is not at a well
defined boundary. Overuse of mocking turns the unit tests into a "change
detector" which will fail whenever the code is modified. Please also consider
whether other testing strategies like injecting fakes/stubs or integration
testing would yield similar test coverage.
```

File: src/gtest/wrappers.h [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+#include "ae.h"
+#include "server.h"
+
+/**
+ * The list of wrapper methods defined.  Each wrapper method must
+ * conform to the same naming conventions (i.e. prefix with a
+ * '__wrap_') and have its method signature match the overridden
+ * method exactly.
+ *
+ * Note: You should NOT wrap variable argument functions (i.e have "...")
```
Comment (murphyjacob4) [inline src/gtest/wrappers.h:56]: Let's add a note here to ensure new usage complies with our guidelines

```suggestion
 * method exactly.
 *
 * Important: Please read the README.md file for guidelines about mocking. Your
 * usage of mocking will not be approved if it doesn't follow the guidelines.
 *
 * Note: You should NOT wrap variable argument functions (i.e have "...")
```

**Review by murphyjacob4 — APPROVED** [review]
Overall LGTM. Still have a few worries about maintenance but I think some additional warnings in the README and wrappers.h will help steer folks in the right direction.

File: deps/Makefile [hunk 21]
```diff
@@ -128,3 +128,13 @@ fast_float_c_interface: .make-prerequisites
...
+
+gtest-parallel: .make-prerequisites
+	@printf '%b %b\n' $(MAKECOLOR)MAKE$(ENDCOLOR) $(BINCOLOR)$@$(ENDCOLOR)
+	@if [ ! -f gtest-parallel/gtest_parallel.py ]; then \
+		echo "Downloading gtest-parallel..."; \
+		rm -rf gtest-parallel; \
+		git clone --depth 1 https://github.com/google/gtest-parallel.git gtest-parallel; \
+	fi
+
+.PHONY: gtest
```
[the remark this answers is not in the document]
Comment (harrylin98) [reply deps/Makefile]: Good catch!

File: deps/gtest-parallel
Comment (zuiderkwast) [reply deps/gtest-parallel:1]: Ahhh, that's a new problem. We added jemalloc and Lua before we started requiring signoff.

Maybe we can skip the signoff for vendored dependencies. @madolson WDYT?

**Comment by harrylin98** [comment]
@zuiderkwast @madolson I squashed the subtree commits into one signed-off commits, it's all good now ;)

**Comment by zuiderkwast** [comment]
> @zuiderkwast @madolson I squashed the subtree commits into one signed-off commits, it's all good now ;)

Then it's not a git subtree anymore. 😄 A subtree contains the git history of that vendored code.

I'm OK with this since it's a small dependency. Please update deps/README and mention there the which exact version we vendored (git hash or tag), so it can traced back. It's good if it's easy to see for example if a future CVE on gtest-parallels affects us or not.

**Comment by harrylin98** [comment]
> > @zuiderkwast @madolson I squashed the subtree commits into one signed-off commits, it's all good now ;)
> 
> Then it's not a git subtree anymore. 😄 A subtree contains the git history of that vendored code.
> 
> I'm OK with this since it's a small dependency. Please update deps/README and mention there the which exact version we vendored (git hash or tag), so it can traced back. It's good if it's easy to see for example if a future CVE on gtest-parallels affects us or not.

Sure, updated deps/README.

File: deps/README.md [hunk 22]
```diff
@@ -121,3 +122,29 @@ To upgrade the library,
...
+
+1. Add the remote and fetch upsteam:
+   ```sh
+   git remote add gtest-parallel https://github.com/google/gtest-parallel.git
+   git fetch gtest-parallel master
+   ```
+
+2. Update the subtree from upstream
+   ```sh
+   git merge --no-ff --allow-unrelated-histories -s subtree -X subtree=deps/gtest-parallel gtest-parallel/master
```
Comment (zuiderkwast) [inline deps/README.md]: This whole section doesn't look right. If we include a gtest-parallels as a single squashed commit, it's not a subtree and this is not the way to update it.... Am I missing anything?

Comment (harrylin98) [reply deps/README.md]: The steps of what I did in current code is:

1. Remove previous deps/gtest-parallel, then commit for the changes (`commit A`)
```
rm -rf deps/gtest-parallel
```

2. Import `gtest-parallel` by using `git subtree add --squash`, this will generate two commits, a squashed upstream commit and a merge commit.
```
git remote add gtest-parallel https://github.com/google/gtest-parallel.git 
git fetch gtest-parallel 
git subtree add --prefix=deps/gtest-parallel gtest-parallel master --squash
```

Because the squashed commit cannot be signed off (even via interactive rebase), so I did a git reset back to `commit A`, kept the subtree changes staged and recommitted them with a proper Signed-off-by


I think this is a clean way to import gtest-parallel without triggering DCO warnings. Let me know if it looks good to you, and I can update the README.md accordingly for future updates in gtest-parallel.

Comment (zuiderkwast) [reply deps/README.md]: Seems good, as long as the DCO check is happy. :)

Comment (harrylin98) [reply deps/README.md]: Cool, just updated README.md. I also follow the steps to update gtest-parallel from an old commit, it works as expected.

File: deps/gtest-parallel
Comment (madolson) [reply deps/gtest-parallel:1]: Yeah, we can force the override of it. Maybe we should have done it as a specific commit, but here we are now.

**Review by madolson — APPROVED** [review]
LGTM. Since this isn't being merged into unstable, just a side branch, will go ahead and merge it and we can do followups as needed there.
